envbuilder.go

```diff
@@ -902,6 +838,340 @@ ENTRYPOINT [%q]`, exePath, exePath, exePath)
 	return nil
 }
 
+// RunCacheProbe performs a 'dry-run' build of the image and checks that
+// all of the resulting layers are present in options.CacheRepo.
+// Logger is the logf to use for all operations.
+// Filesystem is the filesystem to use for all operations.
+// Defaults to the host filesystem.
+func RunCacheProbe(ctx context.Context, opts options.Options) (v1.Image, error) {
+	if !opts.GetCachedImage {
+		return nil, fmt.Errorf("developer error: RunCacheProbe must be run with --get-cached-image")
+	}
+	if opts.CacheRepo == "" {
+		return nil, fmt.Errorf("--cache-repo must be set when using --get-cached-image")
+	}
+
+	stageNumber := 0
+	startStage := func(format string, args ...any) func(format string, args ...any) {
+		now := time.Now()
+		stageNumber++
+		stageNum := stageNumber
+		opts.Logger(log.LevelInfo, "#%d: %s", stageNum, fmt.Sprintf(format, args...))
+
+		return func(format string, args ...any) {
+			opts.Logger(log.LevelInfo, "#%d: %s [%s]", stageNum, fmt.Sprintf(format, args...), time.Since(now))
+		}
+	}
+
+	opts.Logger(log.LevelInfo, "%s - Build development environments from repositories in a container", newColor(color.Bold).Sprintf("envbuilder"))
+
+	caBundle, err := initCABundle(opts.SSLCertBase64)
+	if err != nil {
+		return nil, err
+	}
+
+	cleanupDockerConfigJSON, err := initDockerConfigJSON(opts.DockerConfigBase64)
+	if err != nil {
+		return nil, err
+	}
+	defer func() { _ = cleanupDockerConfigJSON() }() // best effort
+
+	var fallbackErr error
+	var cloned bool
+	if opts.GitURL != "" {
+		endStage := startStage("📦 Cloning %s to %s...",
+			newColor(color.FgCyan).Sprintf(opts.GitURL),
+			newColor(color.FgCyan).Sprintf(opts.WorkspaceFolder),
+		)
+
+		reader, writer := io.Pipe()
+		defer reader.Close()
+		defer writer.Close()
+		go func() {
+			data := make([]byte, 4096)
+			for {
+				read, err := reader.Read(data)
```

Comment:
Nit: consider using https://pkg.go.dev/bufio#NewReaderSize with `ReadBytes(delim byte)` where `delim` is '\r'?

Reply:
pre-existing; will address in follow-up

Reply:
+1